src/core/nginx.go

```diff
@@ -399,81 +399,115 @@ func (n *NginxBinaryType) WriteConfig(config *proto.NginxConfig) (*sdk.ConfigApp
+		if err := ensureFilesAllowed(auxFiles, n.config.AllowedDirectoriesMap, config.GetZaux().GetRootDirectory()); err != nil {
+			return nil, err
+		}
 
-	if len(auxFiles) > 0 {
-		auxPath := config.GetZaux().GetRootDirectory()
-		err = n.env.WriteFiles(configApply, auxFiles, auxPath, n.config.AllowedDirectoriesMap)
+		// TODO: return to Control Plane that there was a rollback
+		err = n.env.WriteFiles(configApply, confFiles, filepath.Dir(details.ConfPath), n.config.AllowedDirectoriesMap)
 		if err != nil {
-			log.Warnf("Auxiliary files write failed: %s", err)
+			log.Warnf("configuration write failed: %s", err)
+			n.writeBackup(config, confFiles, auxFiles)
 			return configApply, err
 		}
-	}
 
-	filesToDelete, ok := generateDeleteFromDirectoryMap(config.DirectoryMap, n.config.AllowedDirectoriesMap)
-	if ok {
-		log.Debugf("use explicit set action for delete files %s", filesToDelete)
-	} else {
-		// Delete files that are not in the directory map
-		filesToDelete = getDirectoryMapDiff(systemNginxConfig.DirectoryMap.Directories, config.DirectoryMap.Directories)
-	}
+		if len(auxFiles) > 0 {
+			auxPath := config.GetZaux().GetRootDirectory()
+			err = n.env.WriteFiles(configApply, auxFiles, auxPath, n.config.AllowedDirectoriesMap)
+			if err != nil {
+				log.Warnf("Auxiliary files write failed: %s", err)
+				return configApply, err
+			}
+		}
 
-	fileDeleted := make(map[string]struct{})
-	for _, file := range filesToDelete {
-		log.Infof("Deleting file: %s", file)
-		if _, ok = fileDeleted[file]; ok {
-			continue
+		filesToDelete, ok := generateDeleteFromDirectoryMap(config.DirectoryMap, n.config.AllowedDirectoriesMap)
+		if ok {
+			log.Debugf("use explicit set action for delete files %s", filesToDelete)
+		} else {
+			// Delete files that are not in the directory map
+			filesToDelete = getDirectoryMapDiff(systemNginxConfig.DirectoryMap.Directories, config.DirectoryMap.Directories)
 		}
 
-		if found, foundErr := FileExists(file); !found {
-			if foundErr == nil {
-				log.Debugf("skip delete for non-existing file: %s", file)
+		fileDeleted := make(map[string]struct{})
+		for _, file := range filesToDelete {
+			log.Infof("Deleting file: %s", file)
+			if _, ok = fileDeleted[file]; ok {
 				continue
 			}
-			// possible perm deny, depends on platform
-			log.Warnf("file exists returned for %s: %s", file, foundErr)
-			return configApply, foundErr
+
+			if err := n.env.DeleteFile(configApply, file); err != nil {
+				return configApply, err
+			}
+
+			fileDeleted[file] = struct{}{}
 		}
-		if err = configApply.MarkAndSave(file); err != nil {
-			return configApply, err
+	} else {
+		configApply, err = sdk.NewConfigApplyWithIgnoreDirectives("", n.config.AllowedDirectoriesMap, n.config.IgnoreDirectives)
+		if err != nil {
+			log.Warnf("config_apply error: %s", err)
+			return nil, err
 		}
-		if err = os.Remove(file); err != nil {
-			return configApply, err
+
+		for _, file := range confFiles {
```

Comment:
instead of two loops, i think you can simplify it with `for _, file := range append(confFiles, auxFiles...)`, I think it looks like both loop logic are the same.

Reply:
Unfortunately I can't as there is one slight difference in the two for loops. The way we determine the `rootDirectoryPath` is different depending on if it's a config or aux file.